File: test/lit/ts/optional.ts
// RUN: %cot emit-cir %s | %FileCheck %s

// libtc: optional type — number | null → !cir.optional<i32>

function getOrNull(x: number): number | null {
    if (x > 0) { return x; }
    return null;
}

// CHECK-LABEL: func.func @getOrNull
// CHECK-SAME: ({{%.*}}: i32) -> !cir.optional<i32>
// CHECK: cir.wrap_optional
// CHECK-SAME: i32 to !cir.optional<i32>
// CHECK: cir.none : !cir.optional<i32>
// CHECK: return

File: test/lit/swift/optional.swift
// RUN: %cot emit-cir %s | %FileCheck %s

// libsc: optional type — Int32?, nil, if let

func getNull() -> Int32 {
    var x: Int32? = nil
    x = 42
    if let val = x {
        return val
    }
    return 0
}

// CHECK-LABEL: func.func @getNull
// CHECK: cir.none : !cir.optional<i32>
// CHECK: cir.store
// CHECK: cir.wrap_optional
// CHECK: cir.store
// CHECK: cir.is_non_null
// CHECK: cir.optional_payload
// CHECK: return
